internal/discovery/discovery.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2023 The Kubernetes Authors All rights reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	"time"
+
+	"k8s.io/client-go/discovery"
+	"k8s.io/klog/v2"
+)
+
+// Interval is the interval at which the discovery client fetches the list of objects.
+const Interval = 30 * time.Second
+
+// Discover starts the discovery process, fetching all the objects that can be listed from the apiserver, every `DiscoveryInterval` seconds.
+func (r *GVKMap) Discover(discoveryClient *discovery.DiscoveryClient) {
+	// TODO: Do not update the cache if the list didn't change since the last fetch.
+	groupList, err := discoveryClient.ServerGroups()
+	if err != nil {
+		klog.Errorln("Failed to fetch server groups:", err)
+	}
+	gvkMap := make(map[string]map[string][]string)
+	for _, group := range groupList.Groups {
+		gvkMap[group.Name] = make(map[string][]string)
+		for _, version := range group.Versions {
+			gvkMap[group.Name][version.Version] = make([]string, 0)
+			resources, err := discoveryClient.ServerResourcesForGroupVersion(version.GroupVersion)
+			if err != nil {
+				klog.Errorln("Failed to fetch server resources for group version:", err)
+			}
+			for _, resource := range resources.APIResources {
+				gvkMap[group.Name][version.Version] = append(gvkMap[group.Name][version.Version], resource.Kind)
+			}
+		}
+	}
+	r.Map = gvkMap
+}
+
+// ResolveGVK resolves the variable VKs to a GVK list.
+func (r *GVKMap) ResolveGVK(gvk GroupVersionKind) (resolvedGVKs []GroupVersionKind) {
+	g := gvk.Group
+	v := gvk.Version
+	k := gvk.Kind
+	if g == "" || g == "*" {
+		klog.Errorln("Group is required in the defined GVK", gvk)
+	}
+	hasVersion := v != "" && v != "*"
+	hasKind := k != "" && k != "*"
+	// No need to resolve, return.
+	if hasVersion && hasKind {
+		return nil
+	}
```

Comment:
returning nil when we have all the information is a bit odd at it would usually mean that we couldn't resolve the GVK. I think it would be better to just return an array of 1 with the GVK